File: c++/0013-roman-to-integer/solution.cpp
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int romanToInt(string s) {
        unordered_map<char, int> digits = {
            {'I', 1},
            {'V', 5},
            {'X', 10},
            {'L', 50},
            {'C', 100},
            {'D', 500},
            {'M', 1000}
        };

        int ans = 0;
        int cur = 0, prev = 0;

        for (int i = s.size() - 1; i >= 0; --i) {
            cur = digits[s[i]];
            ans = (cur < prev) ? (ans - cur) : (ans + cur);
            prev = cur;
        }

        return ans;
    }
};

File: java/0013-roman-to-integer/Solution.java
public class Solution {
    public int romanToInt(String s) {
        int ans = 0;
        int cur = 0, prev = 0;

        for (int i = s.length() - 1; i >= 0; --i) {
            cur = switch (s.charAt(i)) {
                case 'I' -> 1;
                case 'V' -> 5;
                case 'X' -> 10;
                case 'L' -> 50;
                case 'C' -> 100;
                case 'D' -> 500;
                case 'M' -> 1000;
                default -> 0;
            };

            ans = (cur < prev) ? (ans - cur) : (ans + cur);
            prev = cur;
        }

        return ans;
    }
}
